pkg/config/crd/store.go

```diff
@@ -84,40 +85,46 @@ func waitForSynced(ctx context.Context, informers map[string]cache.SharedInforme
 type Store struct {
 	conf         *rest.Config
 	ns           map[string]bool
-	caches       map[string]cache.Store
-	watchCtx     context.Context
-	watchCh      chan store.BackendEvent
 	retryTimeout time.Duration
 
+	cacheMutex sync.Mutex
+	caches     map[string]cache.Store
+
+	watchMutex sync.RWMutex
+	watchCtx   context.Context
+	watchCh    chan store.BackendEvent
+
 	// They are used to inject testing interfaces.
 	discoveryBuilder     func(conf *rest.Config) (discovery.DiscoveryInterface, error)
 	listerWatcherBuilder func(conf *rest.Config) (listerWatcherBuilderInterface, error)
 }
 
 var _ store.Store2Backend = &Store{}
 
-// Init implements store.Store2Backend interface.
-func (s *Store) Init(ctx context.Context, kinds []string) error {
+// checkAndCreateCaches checks the presence of custom resource definitions through the discovery API,
+// and then create caches through lwBUilder which is in kinds. It retries within the timeout duration.
+// If the timeout duration is 0, it waits forever (which should be done within a goroutine).
+// Returns the created shared informers, and the list of kinds which are not created yet.
+func (s *Store) checkAndCreateCaches(
+	ctx context.Context,
+	timeout time.Duration,
+	d discovery.DiscoveryInterface,
+	lwBuilder listerWatcherBuilderInterface,
+	kinds []string) (informers map[string]cache.SharedInformer, remaining []string) {
 	kindsSet := map[string]bool{}
-	for _, kind := range kinds {
-		kindsSet[kind] = true
-	}
-	d, err := s.discoveryBuilder(s.conf)
-	if err != nil {
-		return err
+	for _, k := range kinds {
+		kindsSet[k] = true
 	}
-	lwBuilder, err := s.listerWatcherBuilder(s.conf)
-	if err != nil {
-		return err
-	}
-	s.caches = make(map[string]cache.Store, len(kinds))
-	crdCtx, cancel := context.WithTimeout(ctx, s.retryTimeout)
-	defer cancel()
+	informers = map[string]cache.SharedInformer{}
 	retry := false
-	informers := map[string]cache.SharedInformer{}
-	for len(s.caches) < len(kinds) {
-		if crdCtx.Err() != nil {
-			// TODO: runs goroutines for remaining kinds.
+	retryCtx := ctx
+	if timeout > 0 {
+		var cancel context.CancelFunc
+		retryCtx, cancel = context.WithTimeout(ctx, timeout)
+		defer cancel()
+	}
+	for added := 0; added < len(kinds); {
+		if retryCtx.Err() != nil {
```

Comment:
can you increase to .V(4) otherwise it is printing this message in a tight loop.

`Retrying ...`

Reply:
done